benchmarks/swe_bench/swe_bench.py

```diff
@@ -0,0 +1,277 @@
+"""SWE-bench: Can Language Models Resolve Real-World GitHub Issues?
+
+Carlos E. Jimenez, John Yang, Alexander Wettig, Shunyu Yao, Kexin Pei, Ofir Press, Karthik Narasimhan
+https://arxiv.org/abs/2310.06770
+"""
+
+import json
+import os
+import re
+import shlex
+from pathlib import Path
+from inspect_ai.dataset import Dataset
+
+from docker import DockerClient
+from swebench.harness.constants import (
+    APPLY_PATCH_FAIL,
+    MAP_REPO_TO_INSTALL,
+    MAP_REPO_VERSION_TO_SPECS,
+    RESET_FAILED,
+    TESTS_ERROR,
+    TESTS_TIMEOUT,
+)
+from swebench.harness.log_parsers import MAP_REPO_TO_PARSER
+from swebench.harness.utils import get_test_directives
+
+from inspect_ai import Task, task
+from inspect_ai.dataset import FieldSpec, Sample, hf_dataset
+from inspect_ai.scorer import Score, Scorer, Target, mean, scorer, std
+from inspect_ai.solver import TaskState, use_tools, system_message, generate
+from inspect_ai.tool import bash, python
+from inspect_ai.util import sandbox
+from datasets import load_dataset
+from inspect_ai.solver import Plan, Solver
+
+from typing import Callable
+
+
+INPUT_PROMPT = "Please solve the following issue:\n\n{issue_text}"
+COMPOSE_FILE_DIR = Path(__file__).parent / "resources/compose_files/"
+os.makedirs(COMPOSE_FILE_DIR, exist_ok=True)
+
+SAMPLE_TO_IMAGE_PATH = COMPOSE_FILE_DIR / "sample_to_image.json"
+
+DEFAULT_PLAN = [
+    system_message("Please solve the task below."),
+    use_tools([bash()]),
+    generate(),
+]
+DEFAULT_MAX_MESSAGES = 10
+
+def swe_bench(dataset_name : str = "princeton-nlp/SWE-bench_Verified", split : str = "test", filter :  Callable[Sample,bool] | None = None, plan : Plan | list[Solver] = DEFAULT_PLAN,
+max_messages : int = DEFAULT_MAX_MESSAGES) -> list[Task]:
+
+    dataset = hf_dataset(
+        dataset_name,
+        split=split,
+        sample_fields=FieldSpec(
+            input="problem_statement",
+            id="instance_id",
+            metadata=["base_commit","patch","test_patch","version","repo","environment_setup_commit","PASS_TO_PASS","FAIL_TO_PASS"],
+        )
+    )
+
+    if filter:
+        dataset = dataset.filter(filter)
+
+    return [swe_bench_instance(dataset,sample.id,plan,max_messages) for sample in dataset]
+
+
+@task
+def swe_bench_instance(
```

Comment:
I would like to find a nice way to do two things:

1) Run all (or a subset) of the SWE bench samples as a single eval. For this we would need to allow `sandbox` to vary by `Sample`.

2) Demonstrate how to vary the `plan` in way that plays well with `eval-retry`.

I'll polk at both of these and report back when I have a good solution.